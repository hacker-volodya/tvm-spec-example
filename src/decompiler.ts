import type { Program } from "./core/program";
import { isSingle, isMulti } from "./core/program";
import { defaultPipeline } from "./middle/pipeline";
import { printProgram } from "./backend/printer";
import { liftSliceToIR } from "./frontend/lifter";
import { loadEntrySlice, tryDecodeFunctionDictFromRoot } from "./frontend/loader";

export class Decompiler {
  decompileFile(path: string): Program {
    const root = loadEntrySlice(path);
    const dict = tryDecodeFunctionDictFromRoot(root);
    if (dict) {
      const methods = new Map<number, ReturnType<typeof liftSliceToIR>>();
      dict.forEach((contSlice, id) => {
        const ir = liftSliceToIR(contSlice);
        methods.set(id, this.runMiddle(ir));
      });
      return { kind: 'multi', methods };
    } else {
      const ir = liftSliceToIR(root);
      return { kind: 'single', entry: this.runMiddle(ir) };
    }
  }

  private runMiddle(fn: ReturnType<typeof liftSliceToIR>) {
    const pipeline = defaultPipeline();
    return pipeline.run(fn);
  }

  format(p: Program): string {
    return printProgram(p);
  }
}

