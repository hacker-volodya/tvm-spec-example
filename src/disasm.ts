import { cp0 } from "./tvm-spec";
import { Instruction, Operand } from "./gen/tvm-spec";
import { Bit, Builder, Slice } from "ton3-core";

export type VarMap = { [key: string]: any };

let intToBin = (n: number, size: number) => [...Array(size)].map((x, i) => (n >> i & 1) as Bit).reverse();

let bitsToStr = (bits: Bit[]) => bits.map(x => x.toString()).join('');

let prefixToBin = (prefix: string) => {
    let completionTag = prefix.endsWith('_');
    if (completionTag) {
        prefix.slice(0, -1);
    }
    let bits = prefix.split("").flatMap(hex => intToBin(parseInt(hex, 16), 4));
    if (completionTag) {
        bits = removeCompletionTag(bits);
    }
    return bitsToStr(bits);
};

let removeCompletionTag = (bits: Bit[]) => {
    let newLength = bits.lastIndexOf(1);
    if (newLength == -1) {
        throw new Error("no completion tag");
    }
    return bits.slice(0, newLength);
};

export class OpcodeParser {
    private static _map: Map<string, Instruction>

    private static loadOperand(operand: Operand, slice: Slice) {
        if (operand.type == "uint") {
            return slice.loadUint(operand.size);
        } else if (operand.type == "int") {
            return slice.loadInt(operand.size);
        } else if (operand.type == "ref") {
            return slice.loadRef().slice();
        } else if (operand.type == "pushint_long") {
            return slice.loadInt(8 * slice.loadUint(5) + 19);
        } else if (operand.type == "subslice") {
            let refLength = (operand.refs_add ?? 0) + (operand.refs_length_var_size ? slice.loadUint(operand.refs_length_var_size) : 0);
            let bitLength = (operand.bits_padding ?? 0) + (operand.bits_length_var_size ? slice.loadUint(operand.bits_length_var_size) * 8 : 0);
            let bits = slice.loadBits(bitLength);
            if (operand.completion_tag) {
                bits = removeCompletionTag(bits);
            }
            let refs = slice.refs.slice(0, refLength);
            slice.skipRefs(refLength);
            return new Builder().storeBits(bits).storeRefs(refs).cell().slice();
        } else {
            throw new Error('unimplemented');
        }
    }

    private static getMap() {
        return this._map || (this._map = new Map(cp0.instructions.map(insn => [prefixToBin(insn.bytecode.prefix), insn])))
    }

    private static getLongestPrefixLength() {
        return Math.max(...cp0.instructions.map(insn => prefixToBin(insn.bytecode.prefix).length));
    }

    private static loadPrefix(slice: Slice): Instruction {
        for (let bits = 1; bits <= this.getLongestPrefixLength(); bits++) {
            let prefix = bitsToStr(slice.preloadBits(bits));
            let instruction = this.getMap().get(prefix);
            if (instruction == undefined) continue;
            let rangeCheck = instruction.bytecode.operands_range_check;
            if (rangeCheck != undefined) {
                let operands = new Builder().storeBits(slice.bits).storeRefs(slice.refs).cell().slice().skipBits(prefix.length).loadUint(rangeCheck.length);
                if (operands < rangeCheck.from || operands > rangeCheck.to) {
                    continue;
                }
            }
            slice.skipBits(bits);
            return instruction
        }
        throw new Error("Prefix not found");
    }

    public static nextInstruction(slice: Slice): [Instruction, VarMap] {
        let instruction;
        try {
            instruction = this.loadPrefix(slice);
        } catch (e) {
            throw new Error("OpcodeParser: prefix load error", { cause: e })
        }
        let operands: VarMap = {}
        for (let operand of instruction.bytecode.operands ?? []) {
            try {
                operands[operand.name] = this.loadOperand(operand, slice)
            } catch (e) {
                throw new Error(`OpcodeParser: bad operand ${operand.name} for instruction ${instruction.mnemonic} (${e})`, { cause: e })
            }
        }
        return [instruction, operands]
    }
}