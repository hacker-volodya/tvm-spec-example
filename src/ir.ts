import type { Instruction, PossibleValueTypes } from "./gen/tvm-spec";
import type { Slice, Cell, Hashmap } from "ton3-core";

// Intermediate Representation (IR) to describe decompiled code in a
// transformation-friendly way. Optimizers can operate on this IR and
// a later pretty-printer can render textual output.

export type IRType = PossibleValueTypes[number]; // e.g. 'Integer' | 'Cell' | ...

export type IRValueId = string; // uses existing stack variable ids like 'arg0', 'var3'

export type IRValueRef = {
  id: IRValueId;
  types?: IRType[]; // type hints from spec
};

export type IRValueDef = {
  id: IRValueId;
  types?: IRType[];
};

export type IROperandValue = number | bigint | boolean | Slice | Cell | IRFunction | Map<number, IRFunction> | unknown;

export type IROperands = { [name: string]: IROperandValue };

// Input argument can be a reference to a previously defined value
// or an inline expression (used by optimizations such as inlining).
export type IRInlineExpr = {
  kind: 'inline';
  op: IROpPrim; // inlined op; must be side-effect-free to stay correct
};

export type IRInputArg = IRValueRef | IRInlineExpr;

export type IRInputs = { [name: string]: IRInputArg };

export type IROutputs = { [name: string]: IRValueDef };

export type IROpPrim = {
  kind: 'prim';
  spec: Instruction; // full instruction spec for semantics
  mnemonic: string;   // cached for convenience
  inputs: IRInputs;   // stack inputs consumed by this op
  operands: IROperands; // immediate operands (non-stack)
  outputs: IROutputs; // stack outputs produced by this op
};

export type IRStmt = IROpPrim;

export type IRFunction = {
  kind: 'function';
  // Optional human-friendly name for pretty-printer
  name?: string;
  // Formal parameters (top-of-stack last as in decompiler args order)
  args: IRValueDef[];
  // Linear body (we can attach CFG later if needed)
  body: IRStmt[];
  // Resulting stack values after executing the body
  result: IRValueRef[];
  // Optional tails and errors preserved for diagnostics
  asmTail?: { spec: Instruction; operands: IROperands }[];
  tailSliceInfo?: string; // textual form of tail slice for debug only
  decompileError?: string | null;
  disassembleError?: string | null;
};

export function isIRFunction(x: unknown): x is IRFunction {
  return !!x && typeof x === 'object' && (x as any).kind === 'function';
}

// Minimal IR pretty-printer for debugging and evaluation.
export function formatIR(fn: IRFunction, opts?: { methodId?: number }): string {
  const indent = (s: string, n: number) => s.replace(/^/gm, ' '.repeat(n));
  const fmtTypes = (t?: IRType[]) => t && t.length ? `: ${t.join('|')}` : '';
  const fmtValRef = (v: IRValueRef) => `${v.id}${fmtTypes(v.types)}`;
  const fmtValDef = (v: IRValueDef) => `${v.id}${fmtTypes(v.types)}`;

  // Inlined formatter for continuation operands â€“ full body (debug-friendly)
  const formatInlineFn = (f: IRFunction): string => {
    return formatIR(f);
  };

  const formatInlineMap = (m: Map<number, IRFunction>): string => {
    const entries = [...m.entries()];
    const parts = entries.map(([k, fn]) => `${k}: ${formatInlineFn(fn)}`);
    return `{ ${parts.join(', ')} }`;
  };

  const formatInlineOperand = (v: IROperandValue): string => {
    if (isIRFunction(v)) return formatInlineFn(v);
    if (v instanceof Map) return formatInlineMap(v as Map<number, IRFunction>);
    if (typeof v === 'object' && v !== null) {
      // Slice/Cell or other object: fall back to toString()
      try { return String(v); } catch { return '[object]'; }
    }
    if (typeof v === 'number') return formatNumber(v);
    if (typeof v === 'bigint') return formatBigInt(v);
    return JSON.stringify(v);
  };

  const formatNumber = (n: number): string => {
    // Keep small structural constants as decimals to avoid clutter
    if (Number.isInteger(n) && Math.abs(n) <= 512) return String(n);
    // Otherwise include hex for readability
    const hex = '0x' + (n >>> 0).toString(16);
    return `${n} (${hex})`;
  };

  const formatBigInt = (n: bigint): string => {
    // Show bigint in decimal; append hex if small enough
    const abs = n < 0n ? -n : n;
    if (abs <= 512n) return n.toString();
    const hex = '0x' + abs.toString(16);
    return n.toString() + ' (' + hex + ')';
  };

  const formatInlineOpAsExpr = (st: IROpPrim): string => {
    const insOrder = Object.entries(st.inputs).map(([k, v]) => `${k}=${formatInputArg(v)}`).join(', ');
    const opPairs = Object.entries(st.operands).map(([k, v]) => [k, formatInlineOperand(v)] as [string, string]);
    const hasMultiline = opPairs.some(([, v]) => v.includes('\n'));

    if (!hasMultiline) {
      const ops = opPairs.map(([k, v]) => `${k}=${v}`).join(', ');
      const parts = [insOrder, ops].filter(Boolean).join(' | ');
      return `${st.mnemonic}(${parts})`;
    } else {
      let out = `${st.mnemonic}(` + '\n';
      const indent8 = '        ';
      if (insOrder) {
        out += indent8 + insOrder + (opPairs.length ? ' |' : '') + '\n';
      }
      for (let i = 0; i < opPairs.length; i++) {
        const [k, v] = opPairs[i];
        const lines = v.split('\n');
        out += indent8 + k + '=' + lines[0] + (i < opPairs.length - 1 ? ',' : '') + '\n';
        for (let j = 1; j < lines.length; j++) {
          out += indent8 + ' '.repeat(k.length + 1) + lines[j] + (i < opPairs.length - 1 && j === lines.length - 1 ? ',' : '') + '\n';
        }
      }
      out += `    )`;
      return out;
    }
  };

  const formatInputArg = (a: IRInputArg): string => {
    if ((a as any).kind === 'inline') {
      return formatInlineOpAsExpr((a as IRInlineExpr).op);
    }
    return fmtValRef(a as IRValueRef);
  };
  const argsStr = fn.args.map(a => fmtValDef(a)).join(', ');
  const nameFromMethodId = (id?: number) => {
    if (id === -1) return 'recv_external';
    if (id === 0) return 'recv_internal';
    return undefined;
  };
  const nameStr = fn.name ? ` ${fn.name}` : (opts?.methodId !== undefined ? (nameFromMethodId(opts.methodId) ? ` ${nameFromMethodId(opts.methodId)}` : '') : '');
  const header = opts?.methodId !== undefined ? `/* methodId: ${opts.methodId} */\n` : '';
  let out = header + `function${nameStr} (${argsStr}) {\n`;
  for (const st of fn.body) {
    // Order outputs according to spec value_flow if available
    const getOutputsInOrder = (): IRValueDef[] => {
      const vf: any = (st as any).spec?.value_flow;
      const specStack = vf?.outputs?.stack as any[] | undefined;
      if (!specStack) return Object.values(st.outputs);
      const entries = Object.entries(st.outputs) as [string, IRValueDef][];
      const byName = new Map(entries);
      const constKeys: string[] = entries.filter(([k]) => k.startsWith('const')).map(([k]) => k);
      const used = new Set<string>();
      const ordered: IRValueDef[] = [];
      for (const out of specStack) {
        if (out.type === 'simple') {
          const k = out.name as string;
          const v = byName.get(k);
          if (v) { ordered.push(v as IRValueDef); used.add(k); }
        } else if (out.type === 'const') {
          // consume next const* entry in insertion order
          const k = constKeys.find((x) => !used.has(x));
          if (k) { ordered.push(byName.get(k)!); used.add(k); }
        }
        // conditional outputs are not represented in st.outputs
      }
      // Append any remaining outputs in original insertion order
      for (const [k, v] of entries) {
        if (!used.has(k)) ordered.push(v);
      }
      return ordered;
    };

    const outs = getOutputsInOrder().map(fmtValDef).join(', ');
    const insOrder = Object.entries(st.inputs).map(([k, v]) => `${k}=${formatInputArg(v)}`).join(', ');
    const opPairs = Object.entries(st.operands).map(([k, v]) => [k, formatInlineOperand(v)] as [string, string]);
    const hasMultiline = opPairs.some(([, v]) => v.includes('\n'));

    if (!hasMultiline) {
      const ops = opPairs.map(([k, v]) => `${k}=${v}`).join(', ');
      const parts = [insOrder, ops].filter(Boolean).join(' | ');
      out += `    ${outs ? outs + ' = ' : ''}${st.mnemonic}(${parts});\n`;
    } else {
      out += `    ${outs ? outs + ' = ' : ''}${st.mnemonic}(` + '\n';
      const indent8 = '        ';
      if (insOrder) {
        out += indent8 + insOrder + (opPairs.length ? ' |' : '') + '\n';
      }
      for (let i = 0; i < opPairs.length; i++) {
        const [k, v] = opPairs[i];
        const lines = v.split('\n');
        out += indent8 + k + '=' + lines[0] + (i < opPairs.length - 1 ? ',' : '') + '\n';
        for (let j = 1; j < lines.length; j++) {
          out += indent8 + ' '.repeat(k.length + 1) + lines[j] + (i < opPairs.length - 1 && j === lines.length - 1 ? ',' : '') + '\n';
        }
      }
      out += `    );\n`;
    }
  }
  if (fn.result.length) {
    out += `    // result: ${fn.result.map(fmtValRef).join(', ')}\n`;
  }
  if (fn.decompileError) out += `    // decompilation error: ${fn.decompileError}\n`;
  if (fn.asmTail && fn.asmTail.length) {
    for (const ins of fn.asmTail) {
      const ops = Object.entries(ins.operands).map(([k, v]) => `${k}=${formatInlineOperand(v)}`).join(', ');
      out += `    ${ins.spec.mnemonic} ${ops}\n`;
    }
  }
  if (fn.disassembleError) out += `    // disassemble error: ${fn.disassembleError}\n`;
  if (fn.tailSliceInfo) {
    out += `    // tail slice: ${fn.tailSliceInfo}\n`;
  }
  out += `}`;
  return out;
}
