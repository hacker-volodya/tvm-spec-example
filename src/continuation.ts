import { Cell, Hashmap, Slice } from "ton3-core";
import { OpcodeParser, VarMap } from "./disasm";
import { Instruction } from "./gen/tvm-spec";
import { Stack, StackUnderflowError, StackVariable } from "./stackAnalysis";
import { bitsToIntUint } from "ton3-core/dist/utils/numbers";

const CONTINUATIONS: { [key: string]: string[] } = {
    CALLREF: ["c"],
    JMPREF: ["c"],
    JMPREFDATA: ["c"],
    IFREF: ["c"],
    IFNOTREF: ["c"],
    IFJMPREF: ["c"],
    IFNOTJMPREF: ["c"],
    IFREFELSE: ["c"],
    IFELSEREF: ["c"],
    IFREFELSEREF: ["c1", "c2"],
    IFBITJMPREF: ["c"],
    IFNBITJMPREF: ["c"],
    PUSHCONT: ["s"],
    PUSHCONT_SHORT: ["s"],
}

export class Continuation {
    private args: StackVariable[];
    private code: DecompiledInstruction[];
    private resultStack: StackVariable[];
    private decompileError: any = null;
    private disassembleError: any = null;
    private asmTail: DecodedInstruction[];
    private sliceTail: Slice;

    private constructor(code: DecompiledInstruction[], args: StackVariable[], resultStack: StackVariable[], sliceTail: Slice, asmTail: DecodedInstruction[] = [], decompileError: any = null, disassembleError: any = null) {
        this.args = args;
        this.code = code;
        this.resultStack = resultStack;
        this.asmTail = asmTail;
        this.decompileError = decompileError;
        this.sliceTail = sliceTail;
        this.disassembleError = disassembleError;
    }

    private static dumpOperandConts(operands: VarMap): string[] { 
        let conts = [];
        for (const operand in operands) {
            let v = operands[operand];
            if (v instanceof Continuation) {
                conts.push(v.dump());
                delete operands[operand];
            }
            if (v instanceof Hashmap) {
                v.forEach((methodId: number, cont: Continuation) => {
                    conts.push(`/* methodId: ${methodId} */ ${cont.dump()}`);
                });
                delete operands[operand];
            }
        }
        return conts;
    }

    public dump(): string {
        const indentString = (str: string, count: number, indent = " ") => {
            indent = indent.repeat(count);
            return str.replace(/^/gm, indent);
        };
        const isStackOp = (ins: DecompiledInstruction) => ["stack_basic", "stack_complex"].includes(ins.spec.doc.category);

        // count variable usages
        let useCount: { [name: string]: number } = {};
        for (let instruction of this.code) {
            for (let input of Object.values(instruction.inputs)) {
                let name = (input as any).var.name;
                useCount[name] = (useCount[name] ?? 0) + 1;
            }
        }

        // determine instructions to inline
        let inlineMap: { [name: string]: string } = {};
        let skip = new Set<number>();

        const formatInstruction = (ins: DecompiledInstruction): string => {
            let operandsCopy = { ...ins.operands };
            let conts = Continuation.dumpOperandConts(operandsCopy);
            let operands = Object.values(operandsCopy).map(x => `${x}`);
            let inputVars = Object.values(ins.inputs).map((input: any) => inlineMap[input.var.name] ?? input.var.name);
            let inputStr = operands.concat(...inputVars).concat(...conts).join(', ');
            return `${ins.spec.mnemonic}(${inputStr})`;
        };

        for (let i = 0; i < this.code.length; i++) {
            let instruction = this.code[i];
            if (isStackOp(instruction)) continue;
            let outputs = Object.values(instruction.outputs).map((o: any) => o.var.name);
            if (outputs.length !== 1) continue;
            let outVar = outputs[0];
            if (useCount[outVar] !== 1) continue;
            let j = -1;
            for (let k = i + 1; k < this.code.length; k++) {
                let inputs = Object.values(this.code[k].inputs);
                if (inputs.some((inp: any) => inp.var.name === outVar)) {
                    j = k;
                    break;
                }
            }
            if (j === i + 1 && !isStackOp(this.code[j])) {
                inlineMap[outVar] = formatInstruction(instruction);
                skip.add(i);
            }
        }

        let code = "";
        for (let [index, instruction] of this.code.entries()) {
            if (skip.has(index)) continue;
            // hide stack operations
            if (isStackOp(instruction)) {
                //code += `// ${instruction.spec.mnemonic} ${Object.values(instruction.operands).map(x => `${x}`)}\n`;
                continue;
            }
            let outputVars = Object.values(instruction.outputs).map((output: any) => output.var.name).join(', ');
            let inputVars = Object.values(instruction.inputs).map((input: any) => inlineMap[input.var.name] ?? input.var.name);

            let operandsCopy = { ...instruction.operands };
            let conts = Continuation.dumpOperandConts(operandsCopy);
            let operands = Object.values(operandsCopy).map(x => `${x}`);
            let inputStr = operands.concat(...inputVars).concat(...conts).join(', ');
            code += (outputVars ? `${outputVars} = ` : '') + `${instruction.spec.mnemonic}(${inputStr});\n`;
        }
        code += "// result stack: " + this.resultStack.map(s => s.name).join(', ') + "\n";
        if (this.decompileError) {
            code += "// decompilation error: " + this.decompileError + "\n";
        }
        for (let instruction of this.asmTail) {
            let conts = Continuation.dumpOperandConts(instruction.operands);
            let operands = Object.values(instruction.operands).map(x => `${x}`);
            code += `${instruction.spec.mnemonic} ${operands.concat(...conts).join(', ')}\n`;
        }
        if (this.disassembleError) {
            code += `// Disassemble error: ${this.disassembleError}\n`;
        }
        if (this.sliceTail.bits.length > 0 || this.sliceTail.refs.length > 0) {
            code += `// Tail slice:\n${this.sliceTail}\n`;
        }
        return `function (${this.args.map(a => a.name).join(', ')}) {\n${indentString(code.trimEnd(), 4)}\n}`;
    }

    private static loadOpcode(slice: Slice): DecodedInstruction {
        let [instruction, operands] = OpcodeParser.nextInstruction(slice);
        return {
            spec: instruction,
            operands
        };
    }

    private static decompileInstruction(instruction: DecodedInstruction, stack: Stack): DecompiledInstruction {
        // Handle stack instructions
        if (["stack_basic", "stack_complex"].includes(instruction.spec.doc.category)) {
            stack.execStackInstruction(instruction.spec, instruction.operands);
            return {
                spec: instruction.spec,
                operands: instruction.operands,
                inputs: {},
                outputs: {},
                resultStack: stack.copyEntries()
            };
        }
        // Stop decompilation in case of unknown value flow
        if (instruction.spec.value_flow == undefined || instruction.spec.value_flow.inputs == undefined || instruction.spec.value_flow.outputs == undefined) {
            throw new Error(`instruction is missing value flow: ${instruction.spec.mnemonic}`);
        }
        // Pop inputs
        let stackInputs: VarMap = {};
        if (instruction.spec.value_flow.inputs.stack == undefined) {
            throw new Error(`Unconstrained stack input while parsing ${instruction.spec.mnemonic}`);
        }
        for (let input of instruction.spec.value_flow.inputs.stack.reverse()) {
            if (input.type == 'simple') {
                stackInputs[input.name] = { var: stack.pop(), types: input.value_types };
            } else {
                throw new Error(`not supported stack input '${input.type}' while parsing ${instruction.spec.mnemonic}`);
            }
        }
        // Pop outputs
        let stackOutputs: VarMap = {};
        let constCounter = 0;
        if (instruction.spec.value_flow.outputs.stack == undefined) {
            throw new Error(`Unconstrained stack output while parsing ${instruction.spec.mnemonic}`);
        }
        for (let output of instruction.spec.value_flow.outputs.stack.reverse()) {
            if (output.type == 'simple') {
                stackOutputs[output.name] = { var: stack.push(), types: output.value_types };
            } else if (output.type == 'const') {
                stackOutputs[`const${constCounter++}`] = { var: stack.push(), types: [output.value_type] };
            } else {
                throw new Error(`not supported stack output '${output.type}' while parsing ${instruction.spec.mnemonic}`);
            }
        }
        return {
            spec: instruction.spec,
            operands: instruction.operands,
            inputs: stackInputs,
            outputs: stackOutputs,
            resultStack: stack.copyEntries()
        };
    }

    public static decompile(slice: Slice, initialStack: StackVariable[] = []): Continuation {
        let code: DecompiledInstruction[] = [];
        let stack = new Stack(initialStack);
        let args: StackVariable[] = [];
        let decompileError: any = null;
        let asmTail: DecodedInstruction[] = [];
        while (slice.bits.length > 0) {
            let decodedInsn;
            try {
                decodedInsn = Continuation.loadOpcode(slice);
                if (decodedInsn.spec.mnemonic == "DICTPUSHCONST" && decodedInsn.operands["n"] == 19) {
                    decodedInsn.operands["d"] = Hashmap.parse<Number, Continuation>(19, decodedInsn.operands["d"], {
                        deserializers: {
                            key: k => bitsToIntUint(k, { type: "int" }),
                            value: v => Continuation.decompile(v.slice())
                        }
                    });
                }
                if (CONTINUATIONS[decodedInsn.spec.mnemonic] != undefined) {
                    for (let operand of CONTINUATIONS[decodedInsn.spec.mnemonic]) {
                        decodedInsn.operands[operand] = Continuation.decompile(decodedInsn.operands[operand]);
                    }
                }
            } catch (e) {
                return new Continuation(code, args, stack.copyEntries(), slice, asmTail, decompileError, e);
            }
            if (decompileError == null) {
                for (let t = 0;; t++) {
                    try {
                        console.log("in", stack.dump());
                        console.log(decodedInsn.spec.mnemonic);
                        let stack2 = stack.copy();
                        code.push(Continuation.decompileInstruction(decodedInsn, stack2));
                        stack = stack2;
                        console.log("out", stack.dump());
                        break;
                    } catch (e) {
                        // Check for stack underflow
                        if (e instanceof StackUnderflowError && t < 10) {
                            let newArgs = stack.insertArgs(args.length, e.underflowDepth);
                            args.unshift(...newArgs);
                            continue;
                        }
                        // Stop decompiling, try disassemble tail
                        decompileError = e;
                        break;
                    }
                }
            }
            if (decompileError != null) {
                asmTail.push(decodedInsn);
            }
            // indirect ref jump
            while (slice.bits.length == 0 && slice.refs.length > 0) {
                slice = slice.loadRef().slice();
            }
        }
        return new Continuation(code, args, stack.copyEntries(), slice, asmTail, decompileError);
    }
}

export interface DecodedInstruction {
    spec: Instruction;
    operands: VarMap;
}

export interface DecompiledInstruction {
    spec: Instruction;
    operands: VarMap;
    inputs: VarMap;
    outputs: VarMap;
    resultStack: StackVariable[];
}