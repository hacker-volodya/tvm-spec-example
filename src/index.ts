import { BOC, Hashmap, Builder, Slice } from "ton3-core"
import * as fs from 'fs'
import { Continuation } from "./continuation";
import { formatIR } from "./ir";
import { inlinePrevSingleUse } from "./opt/inline";
import { OpcodeParser } from "./disasm";
import { bitsToIntUint } from "ton3-core/dist/utils/numbers";


const slice = BOC.from(new Uint8Array(fs.readFileSync(process.argv[2]))).root[0].slice();

function cloneSlice(s: Slice): Slice {
    return new Builder().storeBits(s.bits).storeRefs(s.refs).cell().slice();
}

function tryDecodeFunctionDictFromRoot(root: Slice): Map<number, ReturnType<Continuation['toIR']>> | null {
    const sc = cloneSlice(root);
    try {
        // 1) SETCP
        const [ins1] = OpcodeParser.nextInstruction(sc);
        if (ins1.mnemonic !== 'SETCP') return null;
        // 2) DICTPUSHCONST
        const [ins2, ops2] = OpcodeParser.nextInstruction(sc);
        if (ins2.mnemonic !== 'DICTPUSHCONST') return null;
        const n: number = ops2['n'];
        const d: Slice = ops2['d'];
        // 3) DICTIGETJMPZ
        const [ins3] = OpcodeParser.nextInstruction(sc);
        if (ins3.mnemonic !== 'DICTIGETJMPZ') return null;
        // 4) THROWARG
        const [ins4] = OpcodeParser.nextInstruction(sc);
        if (ins4.mnemonic !== 'THROWARG') return null;
        // ensure no extra code in root
        if (sc.bits.length !== 0 || sc.refs.length !== 0) return null;

        // Parse dictionary as functions: key=int, value=Continuation
        const dict = Hashmap.parse<number, ReturnType<typeof Continuation.decompile>>(n, d, {
            deserializers: {
                key: (k: any) => bitsToIntUint(k, { type: 'int' }) as number,
                value: (v: any) => Continuation.decompile(v.slice())
            }
        });

        const result = new Map<number, ReturnType<Continuation['toIR']>>();
        dict.forEach((methodId: number, cont: ReturnType<typeof Continuation.decompile>) => {
            result.set(methodId, cont.toIR());
        });
        return result;
    } catch (_) {
        return null;
    }
}

const dictIR = tryDecodeFunctionDictFromRoot(slice);
if (dictIR) {
    // Print dictionary of functions, sorted by methodId
    const pairs = Array.from(dictIR.entries()).sort((a, b) => a[0] - b[0]);
    for (const [id, fn] of pairs) {
        console.log(`/* methodId: ${id} */`);
        console.log(formatIR(inlinePrevSingleUse(fn)));
        console.log();
    }
} else {
    const cont = Continuation.decompile(slice);
    console.log(formatIR(inlinePrevSingleUse(cont.toIR())))
}
