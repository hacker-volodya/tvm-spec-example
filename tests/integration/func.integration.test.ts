import { Decompiler } from '../../src/decompiler';
import { compileFuncToSlice } from '../helpers/func';

describe('Func → Decompiler → Pseudocode snapshot', () => {
  test('simple arithmetic', async () => {
    const code = await compileFuncToSlice(`
      int add(int a, int b) method_id {
        return a + b;
      }
    `);

    await expect((new Decompiler()).decompileSlice(code)).resolves.toMatchSnapshot();
  });

  test('dicts', async () => {
    const code = await compileFuncToSlice(`
      (slice, int) dict_get?(cell dict, int key_len, slice index) asm(index dict key_len) "DICTGET" "NULLSWAPIFNOT";

      (slice, int) get(cell d, slice key) method_id {
        var (val, success?) = d.dict_get?(8, key);
        return (val, success?);
      }
    `);

    await expect((new Decompiler()).decompileSlice(code)).resolves.toMatchSnapshot();
  });

  test('multi arg use', async () => {
    const code = await compileFuncToSlice(`
      int add(int a, int b) method_id {
        return a + b + a + b;
      }
    `);

    await expect((new Decompiler()).decompileSlice(code)).resolves.toMatchSnapshot();
  });

  test('ifs', async () => {
    const code = await compileFuncToSlice(`
      int add(int a, int b) method_id {
        var x = a - 10;
        var y = b + 20;
        var r = a * b;
        if (a > b) {
          r += a / y;
        } else {
          r += b / x;  
        }
        return r;
      }
    `);

    await expect((new Decompiler()).decompileSlice(code)).resolves.toMatchSnapshot();
  });
});
