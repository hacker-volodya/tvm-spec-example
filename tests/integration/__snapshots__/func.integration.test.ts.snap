// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Func → Decompiler → Pseudocode snapshot dicts 1`] = `
"/* methodId: 0 */
function recv_internal () {
}

/* methodId: 119863 */
function (arg1, arg0) {
    var5, var6 = NULLSWAPIFNOT(x=DICTGET(n=8, D=arg1, k=arg0));
    return var5, var6
}
"
`;

exports[`Func → Decompiler → Pseudocode snapshot ifs 1`] = `
"/* methodId: 0 */
function recv_internal () {
}

/* methodId: 83863 */
function (arg1, arg0) {
    return arg1, arg0, arg1
    // decompilation error: RangeError: Invalid array length
    ADDCONST c=-10
    PUSH i=s1
    ADDCONST c=20
    OVER2 
    MUL 
    PUSH2 i=s4, j=s3
    GREATER 
    PUSHCONT_SHORT s=function (arg4, arg3, arg2, arg1, arg0) {
        return arg0, arg4, arg1
        // decompilation error: RangeError: Invalid array length
        DIV 
        ADD 
    }
    PUSHCONT_SHORT s=function (arg4, arg3, arg2, arg1, arg0) {
        return arg0, arg3, arg2
        // decompilation error: RangeError: Invalid array length
        DIV 
        ADD 
    }
    IFELSE 
}
"
`;

exports[`Func → Decompiler → Pseudocode snapshot multi arg use 1`] = `
"/* methodId: 0 */
function recv_internal () {
}

/* methodId: 83863 */
function (arg1, arg0) {
    var9 = ((arg1 + arg0) + arg1) + arg0;
    return var9
}
"
`;

exports[`Func → Decompiler → Pseudocode snapshot simple arithmetic 1`] = `
"/* methodId: 0 */
function recv_internal () {
}

/* methodId: 83863 */
function (arg1, arg0) {
    var0 = arg1 + arg0;
    return var0
}
"
`;
